src/sqlite3gen.cpp

```diff
@@ -23,77 +23,145 @@
 #include "qtbc.h"
 #include "sqlite3gen.h"
 #include "doxygen.h"
+#include "xmlgen.h"
+#include "xmldocvisitor.h"
 #include "config.h"
 #include "util.h"
+#include "outputlist.h"
 #include "docparser.h"
 #include "language.h"
 
+#include "version.h"
 #include "dot.h"
 #include "arguments.h"
 #include "classlist.h"
 #include "filedef.h"
 #include "namespacedef.h"
 #include "filename.h"
 #include "groupdef.h"
+#include "membername.h"
+#include "memberdef.h"
 #include "pagedef.h"
 #include "dirdef.h"
+#include "section.h"
 
+#include <sys/stat.h>
 #include <qdir.h>
 #include <string.h>
 #include <sqlite3.h>
 
-//#define DBG_CTX(x) printf x
-#define DBG_CTX(x) do { } while(0)
+// enable to show general debug messages
+// #define SQLITE3_DEBUG
 
-const char * schema_queries[][2] = {
+// enable to print all executed SQL statements.
+// I recommend using the smallest possible input list.
+// #define SQLITE3_DEBUG_SQL
+
+# ifdef SQLITE3_DEBUG
+#  define DBG_CTX(x) printf x
+# else // SQLITE3_DEBUG
+#  define DBG_CTX(x) do { } while(0)
+# endif
+
+# ifdef SQLITE3_DEBUG_SQL
+// used by sqlite3_trace in generateSqlite3()
+static void sqlLog(void *dbName, const char *sql){
+  msg("SQL: '%s'\n", sql);
+}
+# endif
+
+const char * table_schema[][2] = {
+  /* TABLES */
+  { "meta",
+    "CREATE TABLE IF NOT EXISTS meta (\n"
+      "\t-- Information about this db and how it was generated.\n"
+      "\t-- Doxygen info\n"
+      "\tdoxygen_version    TEXT PRIMARY KEY NOT NULL,\n"
+      /*
+      Doxygen's version is likely to rollover much faster than the schema, and
+      at least until it becomes a core output format, we might want to make
+      fairly large schema changes even on minor iterations for Doxygen itself.
+      If these tools just track a predefined semver schema version that can
+      iterate independently, it *might* not be as hard to keep them in sync?
+      */
+      "\tschema_version     TEXT NOT NULL, -- Schema-specific semver\n"
+      "\t-- run info\n"
+      "\tgenerated_at       TEXT NOT NULL,\n"
+      "\tgenerated_on       TEXT NOT NULL,\n"
+      "\t-- project info\n"
+      "\tproject_name       TEXT NOT NULL,\n"
+      "\tproject_number     TEXT,\n"
+      "\tproject_brief      TEXT\n"
+    ");"
+  },
   { "includes",
     "CREATE TABLE IF NOT EXISTS includes (\n"
       "\t-- #include relations.\n"
       "\trowid        INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,\n"
       "\tlocal        INTEGER NOT NULL,\n"
-      "\tid_src       INTEGER NOT NULL,  -- File id of the includer.\n"
-      "\tid_dst       INTEGER NOT NULL   -- File id of the includee.\n"
-      ");\n"
-    "CREATE UNIQUE INDEX idx_includes ON includes\n"
-      "\t(local, id_src, id_dst);"
+      "\tsrc_id       INTEGER NOT NULL REFERENCES path, -- File id of the includer.\n"
+      "\tdst_id       INTEGER NOT NULL REFERENCES path, -- File id of the includee.\n"
+      /*
+      In theory we could include name here to be informationally equivalent
```

Comment:
"name" here refers to the inner-text of the includes node. Left it out because the _best_ case is that it duplicates what is in `path.name`, and the worst-case is that it's ambiguous and tricks people into mis-using this association table.